src/writer/to_code.rs

```diff
@@ -128,17 +128,33 @@ impl ToCode for Chunk {
                 ref signal,
                 ref trampoline,
                 in_trait,
+                is_detailed,
             } => {
-                let s1 = format!(
-                    "connect_raw(self.as_ptr() as *mut _, b\"{}\\0\".as_ptr() as *const _,",
-                    signal
-                );
+                let mut v: Vec<String> = Vec::with_capacity(4);
+                if is_detailed {
+                    v.push(format!(
+                        r#"let detailed_signal_name = detail.map(|name| {{ format!("{0}::{{}}\0", name) }});"#,
+                        signal
+                    ));
+                    v.push(format!(
+                        r#"let signal_name_ptr: *const u8 = (if let Some(ref name) = detailed_signal_name {{ name.as_bytes() }} else {{ &b"{0}\0"[..] }}).as_ptr();"#,
```

Comment:
```suggestion
                        r#"let signal_name_ptr = detailed_signal_name.as_ref().map_or(&b"{0}\0"[..], |name| name.as_bytes());"#,
```

Something like that should be a bit more idiomatic in Rust. Then move the `as_ptr()` into the function call to loose the lifetime as late as possible.

Reply:
Heh, at one point in my development I had something similar at that point, then wondered if it was too convoluted. I'll make something like that change, and clarify the lifetime of `detailed_signal_name` as well.

Reply:
That's fair. I always prefer `map_or` over `if let` as it has a nice default case, but always get bitten by the missing `.as_ref()` to not move out of the `Option`.

I don't think it's necessary to "clarify" the lifetime, just as long as `signal_name_ptr` is _not_ a raw pointer. That way it carries the lifetime of `detailed_signal_name` during the remainder of the function scope, disallowing that to be moved or mutated.

Reply:
See what you think with commit 65cb60e50efa.

Reply:
You don't need the explicit drop at the end, `detailed_signal_name` will go out of scope anyway. Rather, my suggestion to solve it this way (by making `signal_name` a `&[u8]` borrow) is to disallow `detailed_signal_name` from being mutably borrowed or moved at the same time. By the way, you don't need to explicitly specify the `&[u8]` type, as shown in the suggestion above.

Reply:
My thought with the drop is that it makes it explicit that `detailed_signal_name` outlasts the call to `connect_raw()`, but if you don't think it needs to be explicit, I can change it back.

Reply:
It doesn't need to be explicit, and by placing `.as_ptr()` inside the call expression (rather than before it) that is already guaranteed.

Reply:
OK, changed in 3d8dc87.

Reply:
Cool, thanks! I'm not a maintainer (can't approve this PR in a sensible way), but they'll likely ask you to squash the patches.

Reply:
Yes, please clean up the commit history a bit :)

Reply:
OK, I'll do that once everyone's done reviewing changes.